### PR TITLE
add secret-manager module with example usage

### PR DESCRIPTION
**Pull request by mateuszwu**

### CODE REVIEW

File: modules/secret-manager/variables.tf [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  description = "Set of Secret Manager Secrets ARNs"
+}
+
+variable "read_access_users" {
+  type        = set(string)
+  description = "Set of users names"
+  default     = []
+}
+
+variable "full_access_users" {
```
Comment (qbart) [inline modules/secret-manager/variables.tf]: let's rename `full_access_users` to `write`, giving  both`read & write` permissions means that user has full access

File: examples/secret-manager/versions.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      source  = "hashicorp/aws"
+      version = "~> 3.0"
+    }
+  }
+  required_version = "~> 1.0.5"
+}
+
+provider "aws" {
+  region = "eu-central-1"
+  default_tags {
```
Comment (qbart) [inline examples/secret-manager/versions.tf]: we can skip tags for examples

File: modules/secret-manager/variables.tf [hunk 3]
```diff
@@ -0,0 +1,16 @@
+variable "secrets" {
```
Comment (qbart) [inline modules/secret-manager/variables.tf:6]: we need to introduce new variable `name_prefix` that will be prepended to resources

File: modules/secret-manager/main.tf [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+resource "aws_iam_policy" "read_policy" {
+  name   = "read_policy"
+  policy = data.aws_iam_policy_document.read_policy.json
+}
+
+resource "aws_iam_policy_attachment" "read_attachment" {
+  name       = "read_attachment"
+  users      = setunion(var.read_access_users, var.full_access_users)
```
Comment (qbart) [inline modules/secret-manager/main.tf]: limit read to only read

File: modules/secret-manager/main.tf [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      "secretsmanager:ListSecretVersionIds",
+      "secretsmanager:PutSecretValue",
+    ]
+
+    resources = var.secrets
+  }
+}
+
+resource "aws_iam_policy" "read_policy" {
+  name   = "read_policy"
```
Comment (qbart) [inline modules/secret-manager/main.tf]: `name = "${var.name_prefix}-secret-read"`

File: modules/secret-manager/main.tf [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    actions = [
+      "secretsmanager:ListSecretVersionIds",
+      "secretsmanager:PutSecretValue",
+    ]
+
+    resources = var.secrets
+  }
+}
+
+resource "aws_iam_policy" "read_policy" {
```
Comment (qbart) [inline modules/secret-manager/main.tf]: it's a policy, not need to call it `read_policy`, it should give a meaning of what it does i.e. `read_secret`

File: modules/secret-manager/main.tf [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    actions = [
+      "secretsmanager:DescribeSecret",
+      "secretsmanager:GetSecretValue",
+    ]
+
+    resources = var.secrets
+  }
+}
+
+data "aws_iam_policy_document" "write_policy" {
```
Comment (qbart) [inline modules/secret-manager/main.tf]: `write_policy` -> `write_secret`

File: modules/secret-manager/main.tf [hunk 8]
```diff
@@ -0,0 +1,51 @@
+data "aws_iam_policy_document" "read_policy" {
```
Comment (qbart) [inline modules/secret-manager/main.tf]:
```suggestion
data "aws_iam_policy_document" "read_secret" {
```

File: modules/secret-manager/main.tf [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    ]
+
+    resources = var.secrets
+  }
+}
+
+data "aws_iam_policy_document" "write_policy" {
+  statement {
+    actions = [
+      "secretsmanager:ListSecretVersionIds",
```
Comment (qbart) [inline modules/secret-manager/main.tf]:
```suggestion
```

Comment (qbart) [reply modules/secret-manager/main.tf]: this is read permission

File: modules/secret-manager/main.tf [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    ]
+
+    resources = ["*"]
+  }
+
+  statement {
+    actions = [
+      "secretsmanager:DescribeSecret",
+      "secretsmanager:GetSecretValue",
+    ]
```
Comment (qbart) [inline modules/secret-manager/main.tf:15]:
```suggestion
      "secretsmanager:ListSecretVersionIds",
    ]
```

File: modules/secret-manager/main.tf (hunk 5 above)
Comment (qbart) [reply modules/secret-manager/main.tf]: IAM names are globally unique, we need this ☝️

File: modules/secret-manager/main.tf [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    resources = var.secrets
+  }
+}
+
+resource "aws_iam_policy" "read_policy" {
+  name   = "read_policy"
+  policy = data.aws_iam_policy_document.read_policy.json
+}
+
+resource "aws_iam_policy_attachment" "read_attachment" {
```
Comment (qbart) [inline modules/secret-manager/main.tf]:
```suggestion
resource "aws_iam_policy_attachment" "secret_read" {
```
attachment is already in resource name

File: modules/secret-manager/main.tf [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  name       = "read_attachment"
```
Comment (qbart) [inline modules/secret-manager/main.tf]:
```suggestion
  name       = "${var.prefix_name}-secret-read"
```

File: modules/secret-manager/main.tf [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  users      = setunion(var.read_access_users, var.full_access_users)
+  policy_arn = aws_iam_policy.read_policy.arn
+}
+
+resource "aws_iam_policy" "write_policy" {
```
Comment (qbart) [inline modules/secret-manager/main.tf]: fix name (similar to read approach)

File: modules/secret-manager/main.tf [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  name   = "write_policy"
```
Comment (qbart) [inline modules/secret-manager/main.tf]:
```suggestion
  name   = "${var.prefix_name}-secret-write"
```

File: modules/secret-manager/main.tf [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  policy = data.aws_iam_policy_document.write_policy.json
+}
+
+resource "aws_iam_policy_attachment" "write_attachment" {
+  name       = "write_attachment"
```
Comment (qbart) [inline modules/secret-manager/main.tf]:
```suggestion
  name       = "${var.prefix_name}-secret-write"
```

File: modules/secret-manager/main.tf [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  users      = setunion(var.read_access_users, var.full_access_users)
+  policy_arn = aws_iam_policy.read_policy.arn
+}
+
+resource "aws_iam_policy" "write_policy" {
+  name   = "write_policy"
+  policy = data.aws_iam_policy_document.write_policy.json
+}
+
+resource "aws_iam_policy_attachment" "write_attachment" {
```
Comment (qbart) [inline modules/secret-manager/main.tf]: fix resource naming

**Review by qbart — CHANGES_REQUESTED** [review]
Few changes needed, I might have copied something wrong, make sure naming is correct

**Comment by qbart** [comment]
@mateuszwu I tried to use it and it fails in the following context (when one of the read/write users field is empty):
```
module "aws_iam_secret_manager_policy" {
  source      = "../../modules/secret-manager"
  name_prefix = "bart-aws-iam-secret-manager"
  secrets     = [aws_secretsmanager_secret.secret_no_1.arn]
  read_users  = []
  write_users = [aws_iam_user.full_access_user.name]
}

module "aws_iam_secret_manager_policy2" {
  source      = "../../modules/secret-manager"
  name_prefix = "bart-aws-iam-secret-manager2"
  secrets     = [aws_secretsmanager_secret.secret_no_2.arn]
  read_users  = [aws_iam_user.read_only_user.name]
  write_users = []
}
```

**Comment by mateuszwu** [comment]
> @mateuszwu I tried to use it and it fails in the following context (when one of the read/write users field is empty):
> 
> ```
> module "aws_iam_secret_manager_policy" {
>   source      = "../../modules/secret-manager"
>   name_prefix = "bart-aws-iam-secret-manager"
>   secrets     = [aws_secretsmanager_secret.secret_no_1.arn]
>   read_users  = []
>   write_users = [aws_iam_user.full_access_user.name]
> }
> 
> module "aws_iam_secret_manager_policy2" {
>   source      = "../../modules/secret-manager"
>   name_prefix = "bart-aws-iam-secret-manager2"
>   secrets     = [aws_secretsmanager_secret.secret_no_2.arn]
>   read_users  = [aws_iam_user.read_only_user.name]
>   write_users = []
> }
> ```

@qbart everything should be working now.

File: examples/secret-manager/main.tf [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+resource "aws_iam_user" "read_only_user" {
+  name = "read_only_user"
+}
+
+resource "aws_iam_access_key" "read_only_user" {
+  user = aws_iam_user.read_only_user.name
+}
+
+output "IAM_user1_id" {
```
Comment (qbart) [inline examples/secret-manager/main.tf]: as a final touch, let's fix upper case style to lower case

**Review by qbart — CHANGES_REQUESTED** [review]
Final touches and we can merge

**Review by qbart — APPROVED** [review]

**Pull request merged**